salt/tests/unit/modules/test_metalk8s.py

```diff
@@ -0,0 +1,272 @@
+
+        patch_dict = {
+            'metalk8s_kubernetes.ping': kubernetes_ping_mock
+        }
+        with patch.dict(metalk8s.__salt__, patch_dict):
+            self.assertEqual(
+                metalk8s.wait_apiserver(retry=retry, interval=0),
+                result
+            )
+
+    @parameterized.expand([
+        (['192.168.1.10'], 'IP:192.168.1.10'),
+        (['fe80::1ff:fe23:4567:890a'], 'IP:fe80::1ff:fe23:4567:890a'),
+        (['192.168.1.10', '192.168.1.11'], 'IP:192.168.1.10, IP:192.168.1.11'),
+        (['my.dns.address'], 'DNS:my.dns.address'),
+        (['my.first.dns', 'my.second.dns'], 'DNS:my.first.dns, DNS:my.second.dns'),
+        (['192.168.1.10', 'my.dns.address'], 'DNS:my.dns.address, IP:192.168.1.10'),
+        (['192.168.1.10', 'my.second.dns', '192.168.1.11', 'my.first.dns'], 'DNS:my.first.dns, DNS:my.second.dns, IP:192.168.1.10, IP:192.168.1.11')
+    ])
+    def test_format_san(self, names, result):
+        """
+        Tests the return of `format_san` function
+        """
+        self.assertEqual(metalk8s.format_san(names), result)
+
+    @parameterized.expand([
+        ({'node1': {'roles': ['master']}, 'node2': {'roles': ['node']}}, 'master', ['node1']),
+        ({'node1': {'roles': ['master', 'node']}, 'node2': {'roles': ['node']}}, 'node', ['node1', 'node2']),
+        ({}, 'master', []),
+        ({'node1': {}, 'node2': {}}, 'master', []),
+        ({'node1': {'roles': ['node']}, 'node2': {'roles': ['node']}}, 'master', []),
+        (None, 'master', ['node1'], False, {'node1': {'roles': ['master']}}),
+        (None, 'master', "Can't retrieve 'metalk8s:nodes' pillar: 'nodes'", True),
+        (None, 'master', "Can't retrieve minions by role because of errors in pillar 'metalk8s:nodes': Error in pillar", True, {'_errors': ['Error in pillar']})
+    ])
+    def test_minions_by_role(self, nodes, role, result,
+                             raises=False, pillar_nodes=None):
+        """
+        Tests the return of `minions_by_role` function
+        """
+        pillar_dict = {
+            'metalk8s': {}
+        }
+        if pillar_nodes:
+            pillar_dict['metalk8s']['nodes'] = pillar_nodes
+
+        with patch.dict(metalk8s.__pillar__, pillar_dict):
+            if raises:
+                self.assertRaisesRegexp(
+                    CommandExecutionError,
+                    result,
+                    metalk8s.minions_by_role,
+                    role,
+                    nodes
+                )
+            else:
+                self.assertEqual(
+                    set(metalk8s.minions_by_role(role, nodes)),
+                    set(result)
+                )
+
+    @parameterized.expand([
+        (PRODUCT_TXT, {'version': '2.5.0', 'name': 'MetalK8s'}),
+        (None, 'Path /my/path/ has no "product.txt"', True),
+        ("Not a good product txt", {'version': None, 'name': None})
+    ])
+    def test_archive_info_from_tree(self, product, result, raises=False):
```

Comment:
I'm don't know either what is the best practices in tests but for this one, we could have real txt file but for `archive_info_from_iso` you will need real ISO and I think it's not really good so here I do exactly the same tests for both function but indeed it could be files.

But I just saw `mock_open` in doc that will be much simpler than what we have right now, I will change it

Reply:
Ok, my point was only for this function, not the ISO one (still we could build an ISO with the same `product.txt` files aforementioned as the first step in the test function..., but we would also need to ensure that we have the binaries to handle the ISO, so I'm not sure it's a good idea).
What you did seems OK to me, just wanted to ask, because as I said I'm not really aware of how things should be done.

Reply:
FWIW, best practice in unit tests is to mock everything outside of the tested function, and to never use real IO (to avoid flakiness / concurrency issues). So it seems like a good approach to me.
It could make sense for us to add such rules in our developer guidelines ;)